workos/utils/_base_http_client.py

```diff
@@ -0,0 +1,191 @@
+import platform
+from typing import (
+    Generic,
+    TypeVar,
+    TypedDict,
+    Union,
+)
+
+import httpx
+
+from workos.exceptions import (
+    ServerException,
+    AuthenticationException,
+    AuthorizationException,
+    NotFoundException,
+    BadRequestException,
+)
+from workos.utils.request import REQUEST_METHOD_DELETE, REQUEST_METHOD_GET
+
+
+_HttpxClientT = TypeVar("_HttpxClientT", bound=Union[httpx.Client, httpx.AsyncClient])
+
+
+DEFAULT_REQUEST_TIMEOUT = 25
+
+
+class PreparedRequest(TypedDict):
+    method: str
+    url: str
+    headers: httpx.Headers
+    params: dict
+    json: dict
+    timeout: int
+
+
+class BaseHTTPClient(Generic[_HttpxClientT]):
+    _client: _HttpxClientT
+    _base_url: str
+    _version: str
+
+    def __init__(
+        self,
+        *,
+        base_url: str,
+        version: str,
+        timeout: int = DEFAULT_REQUEST_TIMEOUT,
+    ) -> None:
+        self.base_url = base_url
+        self._version = version
+        self._timeout = timeout
+
+    def _enforce_trailing_slash(self, url: str) -> str:
+        return url if url.endswith("/") else url + "/"
+
+    def _generate_api_url(self, path: str) -> str:
+        return self._base_url.format(path)
+
+    def _build_headers(
+        self, custom_headers: Union[dict, None], token: str = None
```

Comment:
This is a bit of a Python gotcha. The default value won't be used if we pass the `custom_headers` parameter at all (even if it's None). It's only used when `custom_headers` is not passed.

Reply:
I'm not sure I'm understanding the gotcha.

>  The default value won't be used if we pass the custom_headers parameter at all

...that seems OK? Overriding the default is what I'd expect when passing in a param.

Reply:
We always want it to end up as a dict, not None, before we tack on our base headers. Otherwise will get a NoneType is not subscriptable error.

Reply:
Right, that's why I suggested removing the `None` and just making the type `custom_headers: Dict = {}`

Reply:
Ah, sorry let me give an example. I don't think the type hinting is strong enough to prevent a `None` value disguised as a `dict`:


```python
def test_with_default(value: str = "test"):
    return f"[{value}]"
    
def test_without_default(value: str):
    return f"[{value}]"
    
def test_with_default_in_function(value: str):
    return f"[{value if value is not None else 'test'}]"
    
print(f"When test_with_default is given a value: {test_with_default('provided value')}")
print(f"When test_without_default is given a value: {test_without_default('provided value')}")
print(f"When test_with_default_in_function is given a value: {test_with_default_in_function('provided value')}")

print(f"When test_with_default is given a value kwarg: {test_with_default(value='provided value')}")
print(f"When test_without_default is given a value kwarg: {test_without_default(value='provided value')}")
print(f"When test_with_default_in_function is given a value kwarg: {test_with_default_in_function(value='provided value')}")

print(f"When test_with_default is NOT given a value: {test_with_default()}")
# Throws error: print(f"When test_without_default is NOT given a value: {test_without_default()}")
# Throws error: print(f"When test_with_default_in_function is NOT given a value: {test_with_default_in_function()}")

# This is the case we want to prevent None for. Without the extra check we'd get customer_headers = None since
# we always pass custom_headers:
value = None
print(f"When test_without_default is given a None value: {test_with_default(value=value)}")
print(f"When test_without_default is given a None value: {test_without_default(value=value)}")
print(f"When test_without_default is given a None value: {test_with_default_in_function(value=value)}")
```

Output:
```
When test_with_default is given a value: [provided value]
When test_without_default is given a value: [provided value]
When test_with_default_in_function is given a value: [provided value]
When test_with_default is given a value kwarg: [provided value]
When test_without_default is given a value kwarg: [provided value]
When test_with_default_in_function is given a value kwarg: [provided value]
When test_with_default is NOT given a value: [test]
When test_without_default is given a None value: [None]
When test_without_default is given a None value: [None]
When test_without_default is given a None value: [test]
```

Reply:
And you don't see this in the type hinting?
![CleanShot 2024-07-23 at 11 37 18](https://github.com/user-attachments/assets/e53a3b86-1fa1-4f3e-b293-a5576ec83436)

Reply:
Just a heads up: it's best practice to not use objects as default param values in Python because Python only instantiates them once: https://stackoverflow.com/questions/26320899/why-is-the-empty-dictionary-a-dangerous-default-value-in-python. Strings/ints/other primitives are okay.

It's goofy and annoying to work around. I think your options are
1. Require a dict to be passed in, no default
2. Require Union[dict, None] to be passed in and instantiate if None (which feels like the behavior you want here). You could even default to `None` to make the parameter fully optional, if that's not against the refactor strategy you're using. That's a pattern I've seen in libraries I've used in the past.

Reply:
Oh. Well that's horrible. OK, `None` it is.

But just to be clear, it's only a problem if we mutate that default value or return it such that it could be mutated later, correct?